src/Layout/TaskLayoutOptions.ts

```diff
@@ -1,34 +1,25 @@
-export type TaskLayoutComponent =
+/**
+ * {@link Task} fields used for rendering.
+ *
+ * The order here determines the order that task fields are rendered and written to markdown.
+ */
+export enum TaskLayoutComponent {
     // NEW_TASK_FIELD_EDIT_REQUIRED
-    | 'description'
-    | 'priority'
-    | 'recurrenceRule'
-    | 'createdDate'
-    | 'startDate'
-    | 'scheduledDate'
-    | 'dueDate'
-    | 'doneDate'
-    | 'cancelledDate'
-    | 'blockedBy'
-    | 'id'
-    | 'blockLink';
+    Description = 'description',
```

Comment:
I tried changing this string value to `descriptionxx`, to see what the mode of failure was.

I got these errors:

![image](https://github.com/obsidian-tasks-group/obsidian-tasks/assets/4840096/1d5f9e7d-886e-469c-bc0d-8eb2853efd83)

I'm unsure what the best solution is for this, but I think it's a potential cause of confusion.

Maybe these other locations should also be updated to use the new enums?

Or maybe, at a minimum, the jsdoc here should explicitly say that the enum values must be the names of Task fields?

I believe this diff allows finding all locations that could  maybe be updated to use the new enum...

```diff
Index: src/Layout/TaskLayoutOptions.ts
===================================================================
diff --git a/src/Layout/TaskLayoutOptions.ts b/src/Layout/TaskLayoutOptions.ts
--- a/src/Layout/TaskLayoutOptions.ts	(revision 068d48d8ceb690ba00dbedf92754280123f53e74)
+++ b/src/Layout/TaskLayoutOptions.ts	(date 1706692623167)
@@ -5,18 +5,18 @@
  */
 export enum TaskLayoutComponent {
     // NEW_TASK_FIELD_EDIT_REQUIRED
-    Description = 'description',
-    Id = 'id',
-    BlockedBy = 'blockedBy',
-    Priority = 'priority',
-    RecurrenceRule = 'recurrenceRule',
-    CreatedDate = 'createdDate',
-    StartDate = 'startDate',
-    ScheduledDate = 'scheduledDate',
-    DueDate = 'dueDate',
-    CancelledDate = 'cancelledDate',
-    DoneDate = 'doneDate',
-    BlockLink = 'blockLink',
+    Description = 'descriptionxx',
+    Id = 'idxx',
+    BlockedBy = 'blockedByxx',
+    Priority = 'priorityxx',
+    RecurrenceRule = 'recurrenceRulexx',
+    CreatedDate = 'createdDatexx',
+    StartDate = 'startDatexx',
+    ScheduledDate = 'scheduledDatexx',
+    DueDate = 'dueDatexx',
+    CancelledDate = 'cancelledDatexx',
+    DoneDate = 'doneDatexx',
+    BlockLink = 'blockLinkxx',
 }
 
 export const taskLayoutComponents = Object.values(TaskLayoutComponent);
```

Reply:
This is weird, I did try the same modification, but I get errors only in the `TaskLayoutOptions.ts` O_o

Are you sure that was the only change in your local repository? 
![Снимок экрана 2024-02-03 в 20 31 40](https://github.com/obsidian-tasks-group/obsidian-tasks/assets/93825870/48c97353-d221-4148-942e-b450a4f28c49)

Reply:
Did you do a refactoring, so that all the uses changed to the new `xx` form?

I just edited the string in `TaskLayoutOptions.ts` manually, to see where errors would occur.

I guess another way of expressing what I'm a bit confused by.

How can peopler know, when using `TaskLayoutOptions.ts`, which of these two options to use:

- `TaskLayoutComponent.Description`
- `'description'`

Reply:
Here's another example that could be changed to use the new code:


```diff
Index: src/Renderer/TaskLineRenderer.ts
===================================================================
diff --git a/src/Renderer/TaskLineRenderer.ts b/src/Renderer/TaskLineRenderer.ts
--- a/src/Renderer/TaskLineRenderer.ts	(revision 068d48d8ceb690ba00dbedf92754280123f53e74)
+++ b/src/Renderer/TaskLineRenderer.ts	(date 1706976330284)
@@ -208,7 +208,7 @@
         component: TaskLayoutComponent,
         task: Task,
     ) {
-        if (component === 'description') {
+        if (component === TaskLayoutComponent.Description) {
             componentString = GlobalFilter.getInstance().removeAsWordFromDependingOnSettings(componentString);
 
             const { debugSettings } = getSettings();
@@ -270,7 +270,7 @@
             else return null;
         }
 
-        if (component === 'description') {
+        if (component === TaskLayoutComponent.Description) {
             const tags = internalSpan.getElementsByClassName('tag');
             for (let i = 0; i < tags.length; i++) {
                 const tagName = tags[i].textContent;
```

And another one:

```diff
Index: src/TaskSerializer/DefaultTaskSerializer.ts
===================================================================
diff --git a/src/TaskSerializer/DefaultTaskSerializer.ts b/src/TaskSerializer/DefaultTaskSerializer.ts
--- a/src/TaskSerializer/DefaultTaskSerializer.ts	(revision 068d48d8ceb690ba00dbedf92754280123f53e74)
+++ b/src/TaskSerializer/DefaultTaskSerializer.ts	(date 1706976433281)
@@ -1,5 +1,5 @@
 import type { Moment } from 'moment';
-import { type TaskLayoutComponent, TaskLayoutOptions } from '../Layout/TaskLayoutOptions';
+import { TaskLayoutComponent, TaskLayoutOptions } from '../Layout/TaskLayoutOptions';
 import { Recurrence } from '../Task/Recurrence';
 import { Task } from '../Task/Task';
 import { Priority } from '../Task/Priority';
@@ -136,7 +136,7 @@
 
         switch (component) {
             // NEW_TASK_FIELD_EDIT_REQUIRED
-            case 'description':
+            case TaskLayoutComponent.Description:
                 return task.description;
             case 'priority': {
                 let priority: string = '';
```

Reply:
Many thanks for the details. I didn't user WebStorm rename. But now I tried I still didn't have the errors that you got. I referenced to the code you send and changed the values in the same places. Each time I tried to break the code to ensure that the code is covered by tests so I put `-` to my commits.


> Or maybe, at a minimum, the jsdoc here should explicitly say that the enum values must be the names of Task fields?

Good idea, done.


> Or maybe, at a minimum, the jsdoc here should explicitly say that the enum values must be the names of Task fields?

I think there will be an issue with the code in the code. I think we can later modify the values of the `enum` to accept only `keyof Task` so that there would be no other possibility.

Reply:
Thanks for the edits - I've resolved the conversation